osu.Game/Graphics/Containers/BeatSyncedContainer.cs

```diff
@@ -79,20 +80,31 @@ protected virtual void OnNewBeat(int beatIndex, TimingControlPoint timingPoint,
         }
 
         [Resolved]
-        protected IBindable<WorkingBeatmap> Beatmap { get; private set; }
+        private IBindable<WorkingBeatmap> beatmap { get; set; }
 
         [Resolved(canBeNull: true)]
         protected GameplayClock GameplayClock { get; private set; }
 
+        [Resolved(canBeNull: true)]
+        protected EditorBeatmap EditorBeatmap { get; private set; }
+
+        [Resolved(canBeNull: true)]
+        protected EditorClock EditorClock { get; private set; }
```

Comment:
If I'm honest these changes are kind of dealbreaker tier to me... I don't really want to see this sort of stuff here long term - it feels like setting a very bad precedent. If I were to suggest alternatives:

* Maybe shipping these off to a more isolated place, like an `EditorBeatSyncedContainer` or similar?
* Maybe having some sort of provider interface that can be used to override the game-global stuff?

```csharp
public interface IBeatSyncProvider
{
    ControlPointInfo ControlPoints { get; }
    IClock Clock { get; }
}
```

Reply:
Knew this was coming. I was going to propose a similar solution, but it's a bit more wide-reaching. If required I can do that now, else you can presume it'll be upcoming.

Reply:
I dunno. I get that fixing `BeatSyncedContainer` to work with editor stuff is a potentially wide-reaching refactor that shouldn't happen in this pull, and so I could look away for the time being if there is no other viable way out, but if there's a somewhat local alternative that can work in the interim then I'd prefer that to happen here because I'm really having trouble justifying an approval to myself otherwise.

Reply:
Sure, I'll take a look later.

Keep in mind it already had this stuff for `GameplayClock`, and if I was to fix this I'd want to fix that too. I'd envisage a couple hours of work to standardise this game-wide (ie. have each relevant clock also cache as a beat contributor).

Reply:
Sure, I'm aware, and I'd be fine with half-baking a proper solution here for the time being (i.e. something like: setting up the proper interface/flow but just making it an optional override in this pull, and then making it mandatory and migrating `GameplayClock` across in a follow-up). See how you go I guess.

Reply:
I've made some changes, please check if this is what you expected.

Reply:
Is more or less what I would have liked to see, save for a few small notes below.